src/main/java/com/hierynomus/smbj/io/BufferedInputStreamReader.java

```diff
@@ -31,15 +33,27 @@ public BufferedInputStreamReader(BufferedInputStream inputStream) {
     }
 
     public int read(byte[] byteArray, int offset, int length) throws IOException {
-        int readResult = inputStream.read(byteArray, offset, length);
-        if (readResult == EOF) {
-            isEnd.set(true);
+        if (cachedByte.get() != EOF) {
+            byteArray[offset] = (byte) cachedByte.getAndSet(EOF);
+            int readResult = inputStream.read(byteArray, offset + 1, length - 1);
+            return readResult == EOF ? 1 : (readResult + 1);
+        } else {
+            return inputStream.read(byteArray, offset, length);
         }
-        return readResult;
     }
 
     public boolean isAvailable() {
-        return !isEnd.get();
+        try {
+            if (inputStream.available() > 0 || cachedByte.get() != EOF) {
```

Comment:
I would first check for the cached byte, this is probably always faster, also its easier to read/debug/understand...
```suggestion
            if (cachedByte.get() != EOF) {
                 return true;
            }
            if (inputStream.available() > 0) {
```

Reply:
Branch prediction would help I guess, but I swapped conditions anyway